network-rfc/12_contracts_design.md

```diff
@@ -0,0 +1,90 @@
+ Similar to UniswapV4 Hooks.
+
+### Reward Claims, Exits, and Closure
+
+While the portal is active, SQD providers can claim their proportional share of accumulated rewards at any time by calling the claimRewards function on the PortalProxy, which calculates their share based on their staked balance relative to the total tokens in the portal and transfers the corresponding tokens to them. The portal continues distributing as long as the data consumer injects tokens through the distribute function, with all distributions based on the FeeRouterModule configured splits.
+
+When SQD providers stake their tokens into the portal, they lock them for a minimum duration period. 
+After this minimum lock period expires, providers can request to exit the portal by calling requestExit with their desired withdrawal amount. 
+However, exits are subject to a time-delay mechanism designed to prevent sudden liquidity shocks: 
+the exit delay consists of a base period of 1 epoch plus a percentual delay calculated by the amount being withdrawn. 
+The system allows a maximum of 1% of the total portal liquidity to exit per epoch, meaning if a provider wants to exit 5% of the liquidity, they must wait 1 epoch (base) plus 5 additional epochs (one epoch per 1% of liquidity), totaling 6 epochs before their full withdrawal is processed. 
+For example, if a provider holds 10% of the portal's total SQD and wants to exit their entire position, they would need to wait 1 base epoch + 10 epochs (for the 10% withdrawal) = 11 epochs total. This mechanism ensures gradual exits and maintains portal stability while still allowing providers to eventually withdraw their staked tokens along with any accumulated rewards.
+
+**Importantly, once a provider requests an exit, they stop earning rewards on the requested exit amount during the entire waiting period**
+
+Throughout this entire process, the system maintains upgradeability through the proxy pattern (allowing the factory admin to deploy improved implementations without affecting existing portals), adjustable fee distributions (admins can modify the FeeRouterModule configuration to change allocation percentages), and emergency controls (pausing functionality at both the factory and individual portal levels for security purposes).
+
+
+
+### Exit Delay Formula
+```
+Total Exit Delay = Base Delay + Percentual Delay
+Base Delay = 1 epoch (mandatory for all exits)
+Percentual Delay = (Withdrawal Amount / Total Portal Liquidity) × 100 epochs
+```
+
+
+---
+
+## State Transitions
+
+```
+Collecting → Active → Closed
+     ↓
+  Failed
+```
+
+- **Collecting**: Portal accepting SQD deposits, waiting to reach target before deadline
+- **Active**: Target met, distributing tokens when injected.
+- **Failed**: Deadline passed without reaching target, full refunds enabled
+- **Closed**: Portal closed?
+
+
+### To Discuss
```

Comment:
I think we also talked about support for transferring the stake. Something to add on the next iteration probably, but maybe worth adding to the design doc